### DISCUSSION TITLE
how to add support for optional locale prefix which can be null (default language)

### DISCUSSION THREAD

**Discussion by galexpert** [tighten/ziggy#616] [Q&A]
Hello, help me to set the default locale , but should not be displayed in the url
i also use this extension https://github.com/mcamara/laravel-localization, there are middlewares,
here is my router
```
Route::group([
/* 'prefix' => '{locale?}',*/
'middleware' => [ 'setLocale' ]
], function()
{
Route::get('{locale?}/real-estate/{slug}', [Modules\Realty\Http\Controllers\MainController::class, 'show'])->name('realtyvue.item');
});
```

here is my middleware setLocale

```
public function handle(Request $request, Closure $next)
{
URL::defaults(['locale' => LaravelLocalization::setLocale()]); // en, uk, pl
return $next($request);
}
```

when local == uk
  in vue component
http://domain.loc/uk/real-estate/17-shop Great result!!!! 200 ok

when local == pl
  in vue component
http://domain.loc/pl/real-estate/17-shop Great result!!!! 200 ok

when local == en
I need the default language not to be displayed in the url
in middleware
URL::defaults(['locale' => null]);

  in vue component
http://domain.loc//real-estate/17-shop Wrong bad result!!!! 404 error

and the result should be
  in vue component
http://domain.loc/real-estate/17-shop Great result!!!! 200 ok

Please help me to set up ziggy or laravel correctly when 'locale' must be null

**Comment by bakerkretzmar** [comment]
> in vue component
> http://domain.loc//real-estate/17-shop Wrong bad result!!!! 404 error

Where and how exactly are you generating this specific URL? Is that the URL in the browser or is it a link on the page? Please include the code itself. Thanks.

**Comment by galexpert** [comment]
thank you very much for your reply


i am generating this route in vue component using ziggi library (great product, very helpful!!! thanks a lot)

route() in vue works well with all local (uk,pl,en)
when locale should be displayed in url http://domain.loc/uk/real-estate/17-shop 200ok
and even so http://domain.loc/en/real-estate/17-shop 200ok

but i need to not display the /en/ segment in the url when en=null should be like this http://domain.loc/real-estate/17-shop 200ok
instead the result is this http://domain.loc//real-estate/17-shop 404


help solve this problem!
What am I doing wrong??????
I've already read a whole forum ... and broke my head ...


here is my vue component

snippet where i generate url

```
 <div v-if="items" class="col" v-for="(item, index) of items" :key="item.id">
                <div class="card shadow sm">
                    <div class="thumbnail">
                    <img :src="'/storage/files/' + item.files[0].thumbnails.url" class="bd-placeholder-img " aria-label="Placeholder: Thumbnail" >
                    </div>

                    <h3><a class="link-primary" :href=" route('realtyvue.item', {'slug': item.lang.alias}) " >{{ item.lang.title }}</a ></h3>

                    <div class="card-body"> 
```
	
here is my app.js

```
import { createApp } from 'vue/dist/vue.esm-bundler.js';
import { ZiggyVue } from '/vendor/tightenco/ziggy/dist/vue.es.js';
/*import { Ziggy } from './ziggy.js';*/ // I disabled this, I include the Ziggy configuration file using the directive
......
......
realtyApp.use(ZiggyVue).use(store).mount('#realtyapp');
```

==============
my master.blade  template

```
@routes
@vite(['Modules/Realty/Resources/assets/sass/app.scss', 'Modules/Realty/Resources/assets/js/realtyapp.js'])
```

here is my connected Ziggy

`const Ziggy = {"url":"http:\/\/domain.loc","port":null,"defaults":{"locale":"en"},"routes":{"realtyvue.category" :{"uri":"{locale?}\/realtycat\/{slug}","methods":["GET","HEAD"]},"realtyvue.item":{"uri":"{locale?}\/real-estate\/{slug}","methods":["GET","HEAD"]}}};	
`
I wrote the web.php router code in the post above

I will be very grateful for your help

**Comment by boris-glumpler** [comment]
I just came across this. 

If the route contains an optional parameter like so:

```
'{categories?}/{page:slug}'`
```

and `categories` is not passed to the `route` method:

```js
import routes from './routes.json'

route('page', {page: 'just-some-page'}, undefined, routes)
```

you end up with a URL like this:

```
https://domain.com//just-some-page
```

**Reply by galexpert** [reply]
thanks for your reply... You may be right... when no parameter is passed, it results in an empty segment... I need to solve this problem somehow... I use ziggi along with laravel-localization . It will remove the default locale segment /en/ from the url. And in Vue, I want to get this result: when I need to not display the /en/ segment in the URL, when (en=null ) the result should be like this http://domain.loc/real-estate/17-shop and I get wrong url http://domain.loc//real-estate/17-shop

**Reply by boris-glumpler** [reply]
Well, once you have the URL you can just do a search/replace on it. Something like this here:

```js
import routes from './routes.json'

const chunks = route('page', {page: 'just-some-page'}, undefined, routes).split('://')
const normalizedUrl = chunks[0] + '://' + chunks[1].replaceAll('//', '/')
```

**Reply by galexpert** [reply]
Oh, maybe this will solve my problem... I'll definitely try it... maybe it will help someone else... thanks a lot

**Reply by bakerkretzmar** [reply]
I'm going to take a look at this when I have some time, it might be a bug in Ziggy. What are your `APP_URL` environment variables set to? Do they include a trailing slash?

**Reply by boris-glumpler** [reply]
No trailing slash for me. I prefer to use the actual production URL for development, so it's set to something like this:

```
https://domain.com
```

**Reply by galexpert** [reply]
Thanks for the help, I have the following settings APP_URL=http://domain-rialty.loc

**Reply by bakerkretzmar** [reply]
@boris-glumpler @galexpert how are you using Laravel's PHP `route()` helper function with these routes and parameters?

In my testing it behaves the same way, if you omit an optional parameter in the middle of the route path you end up with an empty segment or double slash. This kind of makes sense, since in `'{categories?}/test/{page:slug}` the optional param is `categories`, the slash is part of the route path just like `test`.

**Reply by boris-glumpler** [reply]
I don't get double slashes. 

```php
route('pages.single', [
     'categories' => null,
     'page'       => 'about',
])
```

The above returns the correct URL for me, so for a route like `{categories?}/test/{page:slug}` I get `https://domain.com/test/about`. Same thing if I omit `categories`. This is on Laravel 10.7.1, btw.

**Reply by bakerkretzmar** [reply]
Sorry the example I shared wasn't actually exactly what I tested with, you're right, in that case it works—but it only works with the _first_ segment, because that part of the path is trimmed from the ends but not inspected inside. Here's my test (let me know if you can't reproduce this):

```php
    /** @test */
    public function optional_params_inside_path()
    {
        app('router')->get('{country?}/test/{language?}/products/{id}', fn () => '')->name('products.show');
        app('router')->getRoutes()->refreshNameLookups();

        $this->assertSame('http://ziggy.dev/ca/test/fr/products/1', route('products.show', ['country' => 'ca', 'language' => 'fr', 'id' => 1]));
        // Optional param in the middle of a path
        $this->assertSame('http://ziggy.dev/ca/test//products/1', route('products.show', ['country' => 'ca', 'id' => 1]));
        // Optional param at the beginning of a path
        $this->assertSame('http://ziggy.dev/test/fr/products/1', route('products.show', ['language' => 'fr', 'id' => 1]));
        // Both
        $this->assertSame('http://ziggy.dev/test//products/1', route('products.show', ['id' => 1]));
    }
```

The second and last assertions reproduce this issue using Laravel's `route()` helper. If anything we would probably match that behaviour, trimming slashes between the host and path but nowhere else. I think that would fix it for both of the examples mentioned in this discussion though.

**Reply by boris-glumpler** [reply]
Yup, these tests all pass! I don't think I ever created a route with 2 optional params, tbh, so at this point I probs won't in the future ;) It'd be great if you decide to match Laravels default behaviour. I think it makes sense, but if you decide against it, it's an easy fix in application code. Cheers!

**Reply by bakerkretzmar** [reply]
Cool, yeah we definitely want to match Laravel so I'll fix this. Whipped up a quick fix here as a proof of concept but I'll put together something cleaner next week: https://github.com/tighten/ziggy/compare/fix-optional-param-extra-slashes

**Reply by galexpert** [reply]
hello, If I understand you correctly... I have to do something like this... in the vue component, I pass the parameter to the route function in this way 

```
<a href="#" class="tag category">Real Estate</a>
                            <a class="title" :href="route('realtyvue.item', { 'locale': null, 'slug': item.lang.alias}) ">{{
                                item.lang.title }}</a>
```

 in my router web.php this route looks like 
```
Route::group([
    'middleware' => [ 'setLocale']
], function()
{ Route::get('{locale?}/real-estate/{slug}', [Modules\Realty\Http\Controllers\MainController::class, 'show'])->name('realtyvue.item') ;
} 
```
middleware works in this route, it sets the current locale

```
$locale = LaravelLocalization::setLocale() == null ? LaravelLocalization::getDefaultLocale() : LaravelLocalization::setLocale();
   //  $locale = LaravelLocalization::setLocale() == null ? 'en' : LaravelLocalization::setLocale();

        URL::defaults(['locale' => $locale]);
        return $next($request);
```

But I still get the double slash http://domain.loc//real-estate/153-fgndvtbv
maybe this problem is because I have Laravel Framework 9.47.0

temporarily I solved this problem in this way... I have a mass 301 redirect for example... http://domain.loc/en/real-estate/153-fgndvtbv -> http://domain.loc/real-estate /153-fgndvtbv this solves the problem in general, but perhaps there is a more elegant solution to the error or a ziggy+laravel setting??? thank you very much for doing this

**Reply by galexpert** [reply]
> Круто, да, мы определенно хотим соответствовать Laravel, так что я это исправлю. В качестве доказательства концепции я набросал здесь быстрое исправление, но на следующей неделе я соберу что-нибудь почище: [https://github.com/tighten/ziggy/compare/fix-Optional-param-extra-slashes .](https://github.com/tighten/ziggy/compare/fix-optional-param-extra-slashes)

Oh, sorry, I didn't see this solution, I'll try it for sure...
